///////////////////////////////////////////////////////////////////////////////////////////////////
// Основные параметры .proto.
///////////////////////////////////////////////////////////////////////////////////////////////////

syntax = "proto3";
package awd.net;

///////////////////////////////////////////////////////////////////////////////////////////////////
// Опции для компиляции .proto в Java.
// Для самой компиляции исп. скрипт "AwdClient/compile_proto_files_for_java.bat".
///////////////////////////////////////////////////////////////////////////////////////////////////

option java_multiple_files = true;
option java_package = "gg.aswedrown.net";

///////////////////////////////////////////////////////////////////////////////////////////////////
// Спецификация протокола AsWeDrown:
// --> https://github.com/AsWeDrown/awd-protocol/wiki/
///////////////////////////////////////////////////////////////////////////////////////////////////

/*
---------------------------------------------------------------------------------------------------
   ПРЕДЫГРОВОЙ ЭТАП (LOBBY STATE)
---------------------------------------------------------------------------------------------------
 */

message CreateLobbyRequest {
  string playerName = 1;
}

/*
---------------------------------------------------------------------------------------------------
   ИГРОВОЙ ЭТАП (PLAY STATE)
---------------------------------------------------------------------------------------------------
 */

message KeepAlive {
  int32 playerId = 1;
  int32 testId   = 2;
}

/*
---------------------------------------------------------------------------------------------------
   ОБЁРТКА ВСЕХ ПАКЕТОВ
   Используется для идентификации пакетов при получении их в виде "сырого" массива байтов.
---------------------------------------------------------------------------------------------------
 */

message PacketWrapper {
  // Номера с    1 по  999 зарезервированы под пакеты предыгрового этапа (LOBBY STATE),
  //        с 1000 по 9999               - под пакеты игрового этапа (PLAY STATE),
  oneof packet {
    // LOBBY STATE
    CreateLobbyRequest createLobbyRequest = 1;

    // PLAY STATE
    KeepAlive keepAlive                   = 1000;
  }
}
